_posts/2021-03-01-introducing-sql-delta-import.md

```diff
@@ -0,0 +1,144 @@
+but no more than 60 concurrent connections will be used for import since max degree of parallelism is 60 (15 executors x 4 cores). 
+`delta.optimizeWrite` and `delta.autoCompact` configuration will yield optimal file size output for the destination table
+
+#### 3.1 `--num-mappers` and data skew just don't play nicely together
+ 
+When `sqoop` imports data, source table will be split into ranges based on `--split-by` column and each mapper 
+would import it's corresponding range. This works good when `--split-by` column has a near uniform distribution 
+of data, but that's not always the case with source tables... As tables age we tend to add additional columns to them to
+take on new business requirements so over time data in latest rows has a higher fill rate than earlier rows. 
+
+![row density increase over time](/post-images/2021-03-sql-delta-import/row_density_increase.png)
+
+Our source tables here at Scribd definitely have these characteristics. We also have some tables that have entire 
+ranges of data missing due to data cleanup. At some point large chunks of data were just deleted from these tables.
+
+![missing rows](/post-images/2021-03-sql-delta-import/missing_rows.png)
+
+This type of data skew will result in processing time skew and output file size skew when you can only control number of 
+mappers. Yes we can introduce additional computed synthetic column in the source table as our `split-by` column but now 
+there is an additional column that does not add business value, app developers need to be aware of it, computing and 
+storing it takes up database resources and if we plan to use it for imports it's better be indexed, thus even more 
+compute and storage resources.
+
+With `sql-delta-import` we can "solve" this problem by making number of chunks much larger than max degree of parallelism.
+This way large chunks with high data density are broken up into smaller pieces that a single executor can handle. 
+Executors that get chunks with little or no data can just quickly process them and move on to do some real work. 
+
+
+### Advanced use cases
+
+For advanced use cases you don't have to use provided spark application directly. `sql-delta-import` 
+libraries can be imported into your own project. You can specify custom data transformations or JDBC dialect to gain a 
+more precised control of data type handling
+
+```scala
+...
+import com.scribd.importer.spark._
+import com.scribd.importer.spark.transform.DataTransform._
+  
+  implicit val spark: SparkSession = SparkSession.builder().master("local").getOrCreate()
+
+
+  // All additional possible jdbc connector properties described here - https://dev.mysql.com/doc/connector-j/8.0/en/connector-j-reference-configuration-properties.html
+  val jdbcUrl = "jdbc:mysql://hostName:port/database"
+
+  val config = ImportConfig(source = "table", destination = "target_database.table", splitBy = "id", chunks = 10)
```

Comment:
Do we have any place where we have define list of properties acceptable to ImportConfig object?

Reply:
https://github.com/delta-io/connectors/pull/80/files#diff-a483fb718651c27ae2ec077a5074e087870e1c013573f8ab3efe2457c5322754R54